### PR TITLE
Add option to the protoc code generator to avoid generating deprecated code

### PR DESCRIPTION
**Pull request by mgodave**
### Motivation

The generated ServiceTalk gRPC stubs make use of deprecated calls. This causes a problem when `-Werror` is used to build the code since it will automatically fail the build. We should allow users who have already migrated their code to prevent the protoc compiler from generating and using deprecated references.

### Modifications

Add an option, `skipDeprecated`, as part of the protoc compiler configuration, to tell the generator to leave out deprecated references.

### Result

Example: Tester.java before/after (option enabled) https://gist.github.com/mgodave/5772193156ca740ea90ca11bafb3da56
Example: Tester.java before/after (option disabled) https://gist.github.com/mgodave/dcff4574f0b70cb39c3bd97a859565ac

Remove
- initSerializationProvider, reason: ContentCodec is deprecated
- isSupportedMessageCodingsEmpty, reason ContentCodec is deprecated
- Deprecated ServiceFactory constructors
- ServiceFactory::Builder references to ContentCodec
- Generated client metadata and associated methods, reason: deprecated

### Testing

Manually tested:
- Add new option to ServiceTalk/examples/grpc/protoc-options and use it to generate test_service.proto which covers all combinations of streaming (or not) services.

### CODE REVIEW

**Comment by mgodave** [comment]
Before and after:

`12d11
< import io.servicetalk.encoding.api.Identity;
15d13
< import io.servicetalk.grpc.api.DefaultGrpcClientMetadata;
20d17
< import io.servicetalk.grpc.api.GrpcClientMetadata;
24d20
< import io.servicetalk.grpc.api.GrpcSerializationProvider;
30d25
< import io.servicetalk.grpc.protobuf.ProtoBufSerializationProviderBuilder;
37d31
< import java.time.Duration;
60,73d53
<     private static GrpcSerializationProvider initSerializationProvider(
<             final List<ContentCodec> supportedMessageCodings) {
<         ProtoBufSerializationProviderBuilder builder = new ProtoBufSerializationProviderBuilder();
<         builder.supportedMessageCodings(supportedMessageCodings);
<         builder.registerMessageType(HelloRequest.class, HelloRequest.parser());
<         builder.registerMessageType(HelloReply.class, HelloReply.parser());
<         return builder.build();
<     }
< 
<     private static boolean isSupportedMessageCodingsEmpty(
<             final List<ContentCodec> supportedMessageCodings) {
<         return supportedMessageCodings.isEmpty() || (supportedMessageCodings.size() == 1 && Identity.identity().equals(supportedMessageCodings.get(0)));
<     }
< 
177,258d156
<         /**
<          * Create a new instance.
<          *
<          * @param service a service to handle incoming requests
<          * @param supportedMessageCodings the set of allowed encodings
<          * @deprecated Use {@link Builder#Builder()}, {@link Builder#bufferDecoderGroup(BufferDecoderGroup)}, and {@link Builder#bufferEncoders(List)}.
<          */
<         @Deprecated
<         public ServiceFactory(final GreeterService service,
<                 final List<ContentCodec> supportedMessageCodings) {
<             this(new Builder(supportedMessageCodings).sayHello(service));
<         }
< 
<         /**
<          * Create a new instance.
<          *
<          * @param service a service to handle incoming requests
<          * @param strategyFactory a factory that creates an execution strategy for different {@link io.servicetalk.router.api.RouteExecutionStrategy#id() id}s
<          * @deprecated Use {@link Builder#Builder()} and set the custom strategy on {@link Builder#routeExecutionStrategyFactory(RouteExecutionStrategyFactory)} instead.
<          */
<         @Deprecated
<         public ServiceFactory(final GreeterService service,
<                 final RouteExecutionStrategyFactory<GrpcExecutionStrategy> strategyFactory) {
<             this(new Builder(strategyFactory).sayHello(service));
<         }
< 
<         /**
<          * Create a new instance.
<          *
<          * @param service a service to handle incoming requests
<          * @param strategyFactory a factory that creates an execution strategy for different {@link io.servicetalk.router.api.RouteExecutionStrategy#id() id}s
<          * @param supportedMessageCodings the set of allowed encodings
<          * @deprecated Use {@link Builder#Builder(RouteExecutionStrategyFactory)}, {@link Builder#bufferDecoderGroup(BufferDecoderGroup)}, and {@link Builder#bufferEncoders(List)}.
<          */
<         @Deprecated
<         public ServiceFactory(final GreeterService service,
<                 final RouteExecutionStrategyFactory<GrpcExecutionStrategy> strategyFactory,
<                 final List<ContentCodec> supportedMessageCodings) {
<             this(new Builder(strategyFactory, supportedMessageCodings).sayHello(service));
<         }
< 
<         /**
<          * Create a new instance.
<          *
<          * @param service a service to handle incoming requests
<          * @param supportedMessageCodings the set of allowed encodings
<          * @deprecated Use {@link Builder#Builder()}, {@link Builder#bufferDecoderGroup(BufferDecoderGroup)}, and {@link Builder#bufferEncoders(List)}.
<          */
<         @Deprecated
<         public ServiceFactory(final BlockingGreeterService service,
<                 final List<ContentCodec> supportedMessageCodings) {
<             this(new Builder(supportedMessageCodings).sayHelloBlocking(service));
<         }
< 
<         /**
<          * Create a new instance.
<          *
<          * @param service a service to handle incoming requests
<          * @param strategyFactory a factory that creates an execution strategy for different {@link io.servicetalk.router.api.RouteExecutionStrategy#id() id}s
<          * @deprecated Use {@link Builder#Builder()} and set the custom strategy on {@link Builder#routeExecutionStrategyFactory(RouteExecutionStrategyFactory)} instead.
<          */
<         @Deprecated
<         public ServiceFactory(final BlockingGreeterService service,
<                 final RouteExecutionStrategyFactory<GrpcExecutionStrategy> strategyFactory) {
<             this(new Builder(strategyFactory).sayHelloBlocking(service));
<         }
< 
<         /**
<          * Create a new instance.
<          *
<          * @param service a service to handle incoming requests
<          * @param strategyFactory a factory that creates an execution strategy for different {@link io.servicetalk.router.api.RouteExecutionStrategy#id() id}s
<          * @param supportedMessageCodings the set of allowed encodings
<          * @deprecated Use {@link Builder#Builder(RouteExecutionStrategyFactory)}, {@link Builder#bufferDecoderGroup(BufferDecoderGroup)}, and {@link Builder#bufferEncoders(List)}.
<          */
<         @Deprecated
<         public ServiceFactory(final BlockingGreeterService service,
<                 final RouteExecutionStrategyFactory<GrpcExecutionStrategy> strategyFactory,
<                 final List<ContentCodec> supportedMessageCodings) {
<             this(new Builder(strategyFactory, supportedMessageCodings).sayHelloBlocking(service));
<         }
< 
264,310d161
<             private final List<ContentCodec> supportedMessageCodings;
< 
<             /**
<              * Create a new instance.
<              */
<             public Builder() {
<                 this(Collections.emptyList());
<             }
< 
<             /**
<              * Create a new instance.
<              *
<              * @param supportedMessageCodings the set of allowed encodings
<              * @deprecated Use {@link #bufferDecoderGroup(BufferDecoderGroup)} and {@link #bufferEncoders(List)}.
<              */
<             @Deprecated
<             public Builder(final List<ContentCodec> supportedMessageCodings) {
<                 this.supportedMessageCodings = supportedMessageCodings;
<             }
< 
<             /**
<              * Create a new instance.
<              *
<              * @param strategyFactory a factory that creates an execution strategy for different {@link io.servicetalk.router.api.RouteExecutionStrategy#id() id}s
<              * @deprecated use {@link #routeExecutionStrategyFactory(RouteExecutionStrategyFactory)} on the Builder instead.
<              */
<             @Deprecated
<             public Builder(
<                     final RouteExecutionStrategyFactory<GrpcExecutionStrategy> strategyFactory) {
<                 this(strategyFactory, Collections.emptyList());
<             }
< 
<             /**
<              * Create a new instance.
<              *
<              * @param strategyFactory a factory that creates an execution strategy for different {@link io.servicetalk.router.api.RouteExecutionStrategy#id() id}s
<              * @param supportedMessageCodings the set of allowed encodings
<              * @deprecated Use {@link #Builder(RouteExecutionStrategyFactory)}, {@link #bufferDecoderGroup(BufferDecoderGroup)}, and {@link #bufferEncoders(List)}.
<              */
<             @Deprecated
<             public Builder(
<                     final RouteExecutionStrategyFactory<GrpcExecutionStrategy> strategyFactory,
<                     final List<ContentCodec> supportedMessageCodings) {
<                 super(strategyFactory);
<                 this.supportedMessageCodings = supportedMessageCodings;
<             }
< 
342,346c193
<                 if (supportedMessageCodings.isEmpty()) {
<                     addRoute(rpc.getClass(), SayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
<                 } else {
<                     addRoute(SayHelloRpc.PATH, rpc.getClass(), SayHelloRpc.methodDescriptor().javaMethodName(), route, HelloRequest.class, HelloReply.class, initSerializationProvider(supportedMessageCodings));
<                 }
---
>                 addRoute(rpc.getClass(), SayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
354,358c201
<                 if (supportedMessageCodings.isEmpty()) {
<                     addRoute(strategy, SayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
<                 } else {
<                     addRoute(SayHelloRpc.PATH, strategy, route, HelloRequest.class, HelloReply.class, initSerializationProvider(supportedMessageCodings));
<                 }
---
>                 addRoute(strategy, SayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
365,369c208
<                 if (supportedMessageCodings.isEmpty()) {
<                     addBlockingRoute(rpc.getClass(), BlockingSayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
<                 } else {
<                     addBlockingRoute(BlockingSayHelloRpc.PATH, rpc.getClass(), BlockingSayHelloRpc.methodDescriptor().javaMethodName(), route, HelloRequest.class, HelloReply.class, initSerializationProvider(supportedMessageCodings));
<                 }
---
>                 addBlockingRoute(rpc.getClass(), BlockingSayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
378,382c217
<                 if (supportedMessageCodings.isEmpty()) {
<                     addBlockingRoute(strategy, BlockingSayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
<                 } else {
<                     addBlockingRoute(BlockingSayHelloRpc.PATH, strategy, route, HelloRequest.class, HelloReply.class, initSerializationProvider(supportedMessageCodings));
<                 }
---
>                 addBlockingRoute(strategy, BlockingSayHelloRpc.methodDescriptor(), bufferDecoderGroup, bufferEncoders, route);
392,403d226
<             /**
<              * Adds a {@link BlockingGreeterService} implementation.
<              *
<              * @param service the {@link BlockingGreeterService} implementation to add.
<              * @return this.
<              * @deprecated Use {@link #addBlockingService(io.grpc.examples.helloworld.Greeter.BlockingGreeterService)}.
<              */
<             @Deprecated
<             public Builder addService(final BlockingGreeterService service) {
<                 return addBlockingService(service);
<             }
< 
443,458d265
<         }
<     }
< 
<     /**
<      * @deprecated This class will be removed in the future in favor of direct usage of {@link GrpcClientMetadata}. Deprecation of {@link GrpcClientMetadata#path()} renders this type unnecessary.
<      */
<     @Deprecated
<     public static final class SayHelloMetadata extends DefaultGrpcClientMetadata {
<         /**
<          * @deprecated Use {@link DefaultGrpcClientMetadata}.
<          */
<         @Deprecated
<         public static final SayHelloMetadata INSTANCE = new SayHelloMetadata();
< 
<         private SayHelloMetadata(final GrpcClientMetadata metadata) {
<             super(SayHelloRpc.PATH, metadata);
460,485d266
< 
<         private SayHelloMetadata() {
<             super(SayHelloRpc.PATH);
<         }
< 
<         public SayHelloMetadata(final GrpcExecutionStrategy strategy) {
<             super(SayHelloRpc.PATH, strategy);
<         }
< 
<         public SayHelloMetadata(final Duration timeout) {
<             super(SayHelloRpc.PATH, timeout);
<         }
< 
<         public SayHelloMetadata(final ContentCodec requestEncoding) {
<             super(SayHelloRpc.PATH, requestEncoding);
<         }
< 
<         public SayHelloMetadata(final GrpcExecutionStrategy strategy,
<                 final ContentCodec requestEncoding) {
<             super(SayHelloRpc.PATH, strategy, requestEncoding);
<         }
< 
<         public SayHelloMetadata(final GrpcExecutionStrategy strategy,
<                 final ContentCodec requestEncoding, final Duration timeout) {
<             super(SayHelloRpc.PATH, strategy, requestEncoding, timeout);
<         }
489,526d269
<         /**
<          * <pre>
<          *  Sends a greeting
<          * </pre>
<          *
<          * @param request the request to send to the server.
<          * @return a {@link Single} which completes when the response is received from the server.
<          */
<         Single<HelloReply> sayHello(HelloRequest request);
< 
<         /**
<          * <pre>
<          *  Sends a greeting
<          * </pre>
<          *
<          * @deprecated Use {@link #sayHello(GrpcClientMetadata,HelloRequest)}.
<          * @param metadata the metadata associated with this client call.
<          * @param request the request to send to the server.
<          * @return a {@link Single} which completes when the response is received from the server.
<          */
<         @Deprecated
<         default Single<HelloReply> sayHello(SayHelloMetadata metadata, HelloRequest request) {
<             return Single.failed(new UnsupportedOperationException("This method is not implemented by " + getClass() + ". Consider migrating to an alternative method or implement this method if it's required temporarily."));
<         }
< 
<         /**
<          * <pre>
<          *  Sends a greeting
<          * </pre>
<          *
<          * @param metadata the metadata associated with this client call.
<          * @param request the request to send to the server.
<          * @return a {@link Single} which completes when the response is received from the server.
<          */
<         default Single<HelloReply> sayHello(GrpcClientMetadata metadata, HelloRequest request) {
<             return sayHello(new SayHelloMetadata(metadata), request);
<         }
< 
534,578d276
<         /**
<          * <pre>
<          *  Sends a greeting
<          * </pre>
<          *
<          * @param request the request from the client.
<          * @return the response from the server.
<          * @throws Exception if an unexpected application error occurs.
<          * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
<          */
<         HelloReply sayHello(HelloRequest request) throws Exception;
< 
<         /**
<          * <pre>
<          *  Sends a greeting
<          * </pre>
<          *
<          * @deprecated Use {@link #sayHello(GrpcClientMetadata,HelloRequest)}.
<          * @param metadata the metadata associated with this client call.
<          * @param request the request from the client.
<          * @return the response from the server.
<          * @throws Exception if an unexpected application error occurs.
<          * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
<          */
<         @Deprecated
<         default HelloReply sayHello(SayHelloMetadata metadata, HelloRequest request) throws
<                 Exception {
<             throw new UnsupportedOperationException("This method is not implemented by " + getClass() + ". Consider migrating to an alternative method or implement this method if it's required temporarily.");
<         }
< 
<         /**
<          * <pre>
<          *  Sends a greeting
<          * </pre>
<          *
<          * @param metadata the metadata associated with this client call.
<          * @param request the request from the client.
<          * @return the response from the server.
<          * @throws Exception if an unexpected application error occurs.
<          * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
<          */
<         default HelloReply sayHello(GrpcClientMetadata metadata, HelloRequest request) throws
<                 Exception {
<             return sayHello(new SayHelloMetadata(metadata), request);
<         }
602,619d299
< 
<         @Override
<         public HelloReply sayHello(final HelloRequest request) throws Exception {
<             return client.sayHello(request).toFuture().get();
<         }
< 
<         @Deprecated
<         @Override
<         public HelloReply sayHello(final SayHelloMetadata metadata, final HelloRequest request)
<                 throws Exception {
<             return client.sayHello(metadata, request).toFuture().get();
<         }
< 
<         @Override
<         public HelloReply sayHello(final GrpcClientMetadata metadata, final HelloRequest request)
<                 throws Exception {
<             return client.sayHello(metadata, request).toFuture().get();
<         }
640,641d319
<             private final GrpcClientCallFactory.ClientCall<HelloRequest, HelloReply> sayHelloCall;
< 
648,652d325
<                 if (supportedMessageCodings.isEmpty()) {
<                     sayHelloCall = factory.newCall(SayHelloRpc.methodDescriptor(), bufferDecoderGroup);
<                 } else {
<                     sayHelloCall = factory.newCall(initSerializationProvider(supportedMessageCodings), HelloRequest.class, HelloReply.class);
<                 }
694,715d366
< 
<             @Override
<             public Single<HelloReply> sayHello(final HelloRequest request) {
<                 return sayHello(supportedMessageCodings.isEmpty() ? new DefaultGrpcClientMetadata() : new SayHelloMetadata(), request);
<             }
< 
<             @Deprecated
<             @Override
<             public Single<HelloReply> sayHello(final SayHelloMetadata metadata,
<                     final HelloRequest request) {
<                 Objects.requireNonNull(metadata);
<                 Objects.requireNonNull(request);
<                 return sayHelloCall.request(metadata, request);
<             }
< 
<             @Override
<             public Single<HelloReply> sayHello(final GrpcClientMetadata metadata,
<                     final HelloRequest request) {
<                 Objects.requireNonNull(metadata);
<                 Objects.requireNonNull(request);
<                 return sayHelloCall.request(metadata, request);
<             }
725,726d375
<             private final GrpcClientCallFactory.BlockingClientCall<HelloRequest, HelloReply> sayHelloCall;
< 
733,737d381
<                 if (supportedMessageCodings.isEmpty()) {
<                     sayHelloCall = factory.newBlockingCall(BlockingSayHelloRpc.methodDescriptor(), bufferDecoderGroup);
<                 } else {
<                     sayHelloCall = factory.newBlockingCall(initSerializationProvider(supportedMessageCodings), HelloRequest.class, HelloReply.class);
<                 }
759,780d402
< 
<             @Override
<             public HelloReply sayHello(final HelloRequest request) throws Exception {
<                 return sayHello(supportedMessageCodings.isEmpty() ? new DefaultGrpcClientMetadata() : new SayHelloMetadata(), request);
<             }
< 
<             @Deprecated
<             @Override
<             public HelloReply sayHello(final SayHelloMetadata metadata, final HelloRequest request)
<                     throws Exception {
<                 Objects.requireNonNull(metadata);
<                 Objects.requireNonNull(request);
<                 return sayHelloCall.request(metadata, request);
<             }
< 
<             @Override
<             public HelloReply sayHello(final GrpcClientMetadata metadata,
<                     final HelloRequest request) throws Exception {
<                 Objects.requireNonNull(metadata);
<                 Objects.requireNonNull(request);
<                 return sayHelloCall.request(metadata, request);
<             }
`

**Review by bryce-anderson — APPROVED** [review]
Token approval: this looks fine to me but please wait for a more experienced reviewer before merging.

How was it tested? I presume manually, but it would be good to document it in the PR description.

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java [hunk 1]
```diff
@@ -248,13 +248,15 @@ private State(ServiceDescriptorProto serviceProto, GenerationContext context, St
     private final Map<String, ClassName> messageTypesMap;
     private final ServiceCommentsMap serviceCommentsMap;
     private final boolean printJavaDocs;
+    private final boolean skipDeprecated;
 
     Generator(final GenerationContext context, final Map<String, ClassName> messageTypesMap,
-              final boolean printJavaDocs, SourceCodeInfo sourceCodeInfo) {
+              final boolean printJavaDocs, final boolean skipDeprecated, SourceCodeInfo sourceCodeInfo) {
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:254]: Overall approach lgtm, but I added new option to our protoc-options example, run `./gradlew :servicetalk-examples:grpc:servicetalk-examples-grpc-protoc-options:build`, and this is what I found: 

1. Client interfaces are empty now:

before:
```java
    public interface GreeterClient extends GrpcClient<BlockingGreeterClient> {
        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param request the request to send to the server.
         * @return a {@link Single} which completes when the response is received from the server.
         */
        Single<HelloReply> sayHello(HelloRequest request);

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @deprecated Use {@link #sayHello(GrpcClientMetadata,HelloRequest)}.
         * @param metadata the metadata associated with this client call.
         * @param request the request to send to the server.
         * @return a {@link Single} which completes when the response is received from the server.
         */
        @Deprecated
        default Single<HelloReply> sayHello(SayHelloMetadata metadata, HelloRequest request) {
            return Single.failed(new UnsupportedOperationException("This method is not implemented by " + getClass() + ". Consider migrating to an alternative method or implement this method if it's required temporarily."));
        }

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param metadata the metadata associated with this client call.
         * @param request the request to send to the server.
         * @return a {@link Single} which completes when the response is received from the server.
         */
        default Single<HelloReply> sayHello(GrpcClientMetadata metadata, HelloRequest request) {
            return sayHello(new SayHelloMetadata(metadata), request);
        }

        @Override
        default BlockingGreeterClient asBlockingClient() {
            return new GreeterClientToBlockingGreeterClient(this);
        }
    }

    public interface BlockingGreeterClient extends BlockingGrpcClient<GreeterClient> {
        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param request the request from the client.
         * @return the response from the server.
         * @throws Exception if an unexpected application error occurs.
         * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
         */
        HelloReply sayHello(HelloRequest request) throws Exception;

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @deprecated Use {@link #sayHello(GrpcClientMetadata,HelloRequest)}.
         * @param metadata the metadata associated with this client call.
         * @param request the request from the client.
         * @return the response from the server.
         * @throws Exception if an unexpected application error occurs.
         * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
         */
        @Deprecated
        default HelloReply sayHello(SayHelloMetadata metadata, HelloRequest request) throws
                Exception {
            throw new UnsupportedOperationException("This method is not implemented by " + getClass() + ". Consider migrating to an alternative method or implement this method if it's required temporarily.");
        }

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param metadata the metadata associated with this client call.
         * @param request the request from the client.
         * @return the response from the server.
         * @throws Exception if an unexpected application error occurs.
         * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
         */
        default HelloReply sayHello(GrpcClientMetadata metadata, HelloRequest request) throws
                Exception {
            return sayHello(new SayHelloMetadata(metadata), request);
        }
    }
```

after:
```java
    public interface GreeterClient extends GrpcClient<BlockingGreeterClient> {
        @Override
        default BlockingGreeterClient asBlockingClient() {
            return new GreeterClientToBlockingGreeterClient(this);
        }
    }

    public interface BlockingGreeterClient extends BlockingGrpcClient<GreeterClient> {
    }
```

Expectation: they should have 2 methods: `sayHello(HelloRequest)` and `sayHello(GrpcClientMetadata metadata, HelloRequest request)`. The last one should not have `default` implementation anymore, bcz that `default` was added for backward compatibility only.

2. There are still some fields and methods with `@Deprecated` annotation:

```java
    @FunctionalInterface
    public interface BlockingSayHelloRpc extends BlockingGrpcService {
        /**
         * @deprecated Use {@link #methodDescriptor}.
         */
        @Deprecated
        String PATH = "/helloworld.Greeter/SayHello";
```

```java
            @Override
            @Deprecated
            protected GreeterServiceFromRoutes newServiceFromRoutes(
                    final GrpcRoutes.AllGrpcRoutes routes) {
                return new GreeterServiceFromRoutes(routes);
            }
```

Consider following the same approach for tests. However, existing `helloworld.proto` used by `protoc-options` has only one endpoint (aggregated). Only for the purpose of local testing, consider also copying `test_service.proto` into `protoc-options`. That will codegen all API variations we have. Would be better to use for `skipDeprecated=false|true` diff comparison.

**Review by idelpivnitskiy — CHANGES_REQUESTED** [review]

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java (hunk 1 above)
Comment (mgodave) [reply servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:254]: Fixed in the latest diff.

**Comment by mgodave** [comment]
New diff from test_service.proto: https://gist.github.com/mgodave/5772193156ca740ea90ca11bafb3da56

**Review by idelpivnitskiy — COMMENTED** [review]
Really good progress! New generated code looks much better than without `skipDeprecated` option.
Just a few adjustments for the final findings and we are ready to ship:

File: servicetalk-examples/grpc/protoc-options/src/main/proto/test_service.proto [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "io.servicetalk.grpc.netty";
+option java_outer_classname = "TesterProto";
+
+package grpc.netty;
+
+service Tester {
```
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-examples/grpc/protoc-options/src/main/proto/test_service.proto:24]: If it was handy for testing I feel like we should leave it or add another project just for this. I'm happy to add a usage here but it seems kind of counterproductive to have to copy/paste something in here every time we want to generate a comprehensive example for testing/comparison.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java [hunk 3]
```diff
@@ -1341,10 +1347,8 @@ void handle(GrpcServiceContext ctx, BlockingIterable<Req> request,  // FIXME: 0.
          * The implementation of this method is responsible for calling {@link GrpcPayloadWriter#close()}.
          * @throws Exception If an exception occurs during request processing.
          */
-        default void handle(GrpcServiceContext ctx, BlockingIterable<Req> request,  // FIXME: 0.43 - remove default impl
-                    BlockingStreamingGrpcServerResponse<Resp> response) throws Exception {
-            handle(ctx, request, response.sendMetaData());
-        }
+        void handle(GrpcServiceContext ctx, BlockingIterable<Req> request,
+                    BlockingStreamingGrpcServerResponse<Resp> response) throws Exception;
```
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java:1351]: making this `default` will break the `@FunctionalInterface` contract.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java [hunk 4]
```diff
@@ -1496,10 +1503,8 @@ protected interface BlockingResponseStreamingRoute<Req, Resp>
          * The implementation of this method is responsible for calling {@link GrpcPayloadWriter#close()}.
          * @throws Exception If an exception occurs during request processing.
          */
-        default void handle(GrpcServiceContext ctx, Req request,  // FIXME: 0.43 - remove default impl
-                    BlockingStreamingGrpcServerResponse<Resp> response) throws Exception {
-            handle(ctx, request, response.sendMetaData());
-        }
+        void handle(GrpcServiceContext ctx, Req request,  // FIXME: 0.43 - remove default impl
```
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java:1506]: same as above, re: `@FunctionalInterface`

File: servicetalk-examples/grpc/protoc-options/src/main/proto/test_service.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/grpc/protoc-options/src/main/proto/test_service.proto:24]: You can remove it at the end before merging PR. It just doesn't fit into this example. And creating another module is a bit out of scope of this PR, we will plan that for the follow-up.

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java [hunk 5]
```diff
@@ -847,76 +787,202 @@ supportedMessageCodings, serviceFactoryBuilderInitChain(state.serviceProto, true
...
+                                .addStatement("super($T.$L, $L)", rpcInterface.className, PATH, strategy)
+                                .build())
+                        .addMethod(constructorBuilder()
+                                .addModifiers(PUBLIC)
+                                .addParameter(Duration.class, timeout, FINAL)
+                                .addStatement("super($T.$L, $L)", rpcInterface.className, PATH, timeout)
+                                .build())
+                        .addMethod(constructorBuilder()
+                                .addModifiers(PUBLIC)
+                                .addParameter(ContentCodec, requestEncoding, FINAL)
```
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:963]: The main proto generator still generates calls to the constructors with `supportedMessageCodings`.

Ex: TesterProto.java

```
public static class ClientFactory extends io.servicetalk.grpc.api.GrpcClientFactory<io.servicetalk.grpc.netty.TesterProto.Tester.TesterClient, io.servicetalk.grpc.netty.TesterProto.Tester.BlockingTesterClient> {
      @java.lang.Override
      protected io.servicetalk.grpc.netty.TesterProto.Tester.TesterClient newClient(
          final io.servicetalk.grpc.api.GrpcClientCallFactory factory) {
        return new io.servicetalk.grpc.netty.TesterProto.Tester.ClientFactory.DefaultTesterClient(factory, supportedMessageCodings(), bufferDecoderGroup());
      }
```

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java:1351]: Then unfortunately we won't be able to make this change bcz it's breaking API. If we still override deprecated method in the generated code but don't add `@Deprecated` annotation`, will it pass javac linter?

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java [hunk 6]
```diff
@@ -687,16 +737,22 @@ private TypeSpec.Builder addServiceFactory(final State state, final TypeSpec.Bui
         state.serviceProto.getMethodList().stream()
                 .map(Generator::routeName)
                 .forEach(n -> {
-                    registerRoutesMethodSpecBuilder.addStatement("$L($L)", n, service);
+                    if (!skipDeprecated) {
+                        registerRoutesMethodSpecBuilder.addStatement("$L($L)", n, service);
```
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:741]: good catch!

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:963]: I'm looking at classes generated with `java_multiple_files = true` that are outside of `TestProto` file. Those ctros are generated by our codegen

File: servicetalk-examples/grpc/protoc-options/src/main/proto/test_service.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-examples/grpc/protoc-options/src/main/proto/test_service.proto:24]: sure

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java:1351]: I believe as long as we don't call the method anywhere, even with `@Deprecated`, we will pass the linter. My initial commit had taken this approach and just removed code paths that had our generated code making calls to deprecated code. I'll of course test this out again here.

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgodave) [reply servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:963]: I got turned around by the compiler errors and was looking somewhere else.

File: servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java [hunk 7]
```diff
@@ -1486,36 +1660,48 @@ private static ClassName routeInterfaceClass(final MethodDescriptorProto methodP
...
+                            request)
+                    .addParameter(ParameterizedTypeName.get(GrpcPayloadWriter, outClass),
+                            responseWriter);
+
+            if (!skipDeprecated) {
+                handleMethodSpecBuilder.addStatement("$L.$L($L, $L, $L)",
+                        rpc, routeName, ctx, request, responseWriter);
+            } else {
+                handleMethodSpecBuilder.addStatement("new UnsupportedOperationException(\"" +
+                        "This method is deprecated and should not be called by generated code. \")");
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java:1688]:
```suggestion
                handleMethodSpecBuilder.addStatement("throw new UnsupportedOperationException(\"" +
                        "This method is deprecated and should not be called by router or any generated code.\")");
```

**Review by idelpivnitskiy — APPROVED** [review]
Great work!

The generated code with `skipDeprecated=true` looks good. I also checked diff for default behavior. Many things changed their order, which is ok. Couldn't spot anything suspicious, but would be great if someone can have another look at the diff @bryce-anderson.

**Review by bryce-anderson — APPROVED** [review]
I also looked at a diff of the generated code and didn't see anything suspicious. That said, it's a lot of generated code. 😅

**Pull request merged**